aarch64: use fixed virtual address for kernel address space

midnightveil opened the pull request:
This matches RISC-V which also does this (edit: actually, it doesn't, even though it maps the kernel ELF twice)

This is a precursor to some other boot protocol changes as a part of
the multikernel work where it is useful to be able to load the kernel
at different physical addresses using the same kernel image.

Previously, the kernel would run in the kernel physical memory window
at a platform-specific virtual address.

Indanz reviewed (commented):
Please focus this PR on the fixed virtual address change and do the cleanups and improvements in a separate PR. Small, trivial changes are okay if they're combined, but they need to be relevant.

Updating the diagram is a border case, I think doing that in this PR is okay because you need to update it for this change anyway and it was wrong and misleading. But please split it up in two parts: One that updates it for this PR, and one that does the generic improvements. The order doesn't matter and the update necessary because of commit 58a2 is best done as part of that commit.

midnightveil commented on include/machine.h:
```diff
@@ -13,7 +13,12 @@
+ *
+ * As a result, this function should _not_ be used to modify memory if the
+ * addresses might lie within the kernel ELF mapping.
+ * See the comments in `map_kernel_elf_image()` on AArch64.
```
Yes, but I'm being conservative.

I think I can add an assert here, I'll double check. I had the feeling it would have been hard but it should be doable...

midnightveil commented on include/arch/arm/arch/64/mode/hardware.h:
```diff
@@ -11,155 +11,222 @@
+ *    In EL2 there is no canonical-high portion of the address space since
+ *    address tagging is not supported in EL2. Therefore the kernel is linked
+ *    to the canonical lower portion of the address space (all the unused top bits
+ *    are set to 0, not 1). However, we get a separate set of page tables.
```
This is unmodified comments (I didn't adjust HYP) but I can update this.

midnightveil commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+     *  match that of the kernel ELF mapping, so this is OK. At the same time,
+     *  the current physical window mapping does use NORMAL memory covering memory
+     *  that is not DRAM, which is questionable for newer processors with more
+     *  speculative execution.
```
Yes.

midnightveil commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
-                                                                                         KLVL_FRM_ARM_PT_LVL(2))] = pte_pte_page_new(
+BOOT_CODE void map_kernel_elf_image(const word_t kernel_mappings_pud_idx)
+{
+    compile_assert(elf_paddr_aligned, IS_ALIGNED(KERNEL_ELF_PADDR_BASE_RAW, seL4_LargePageBits));
```
Fixed now by adding alignment.

midnightveil replied:
Although perhaps that's not the correct solution, since the issue isn't that the physBase needs to be aligned but that the kernelElfBase does ..

midnightveil commented on include/arch/arm/arch/64/mode/hardware.h:
```diff
@@ -11,155 +11,222 @@
+ *       up to the second-to-last PUD entry. This covers 255GiB.
+ *
+ *       The physical memory window is mapped using Large Pages.
+         FIXME: Why not huge pages?
```
https://github.com/seL4/seL4/pull/1523/files#r2438356437

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+    compile_assert(elf_base_aligned, IS_ALIGNED(KERNEL_ELF_BASE, seL4_LargePageBits));
+    /* asserts because ELF_TOP is dynamic */
+    assert(GET_KPT_INDEX(KERNEL_ELF_BASE, KLVL_FRM_ARM_PT_LVL(1)) == kernel_mappings_pud_idx);
+    assert(GET_KPT_INDEX(KERNEL_ELF_TOP, KLVL_FRM_ARM_PT_LVL(1)) == kernel_mappings_pud_idx);
```
Considering you know what it should be, wouldn't it make more sense to move the `armKSGlobalKernelPUD[kernel_mappings_pud_idx] = pte_pte_table_new(addrFromKPPtr(&armKSGlobalKernelPDs[kernel_mappings_pud_idx][0]));` here and remove the function argument?

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+
+    word_t paddr = KERNEL_ELF_PADDR_BASE;
+    word_t vaddr = KERNEL_ELF_BASE;
+    for (word_t pd_idx = kernel_elf_pd_start; pd_idx < kernel_elf_pd_end; pd_idx++) {
```
`i` instead of `pd_idx` would be clearer.

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+                  /* AttrIndx */ NORMAL
+              );
+
+        assert(pd_idx == GET_KPT_INDEX(vaddr, KLVL_FRM_ARM_PT_LVL(2)));
```
You already checked the first index with above, I would move this one after the increments. Then you always check things before doing something.

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+
+    /* put the PD into the PUD for kernel mappings */
+    armKSGlobalKernelPUD[kernel_mappings_pud_idx]
+        = pte_pte_table_new(addrFromKPPtr(&armKSGlobalKernelPDs[kernel_mappings_pud_idx][0]));
```
Move this into `map_kernel_elf_image()`.

midnightveil commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
-                                                                               addrFromKPPtr(armKSGlobalKernelPT)
-                                                                           );
+    armKSGlobalKernelPDs[kernel_mappings_pud_idx][device_mapping_pd_idx]
+        = pte_pte_table_new(addrFromKPPtr(armKSGlobalKernelPT));
```
Yes, I wish that was possible too.

midnightveil commented on include/arch/arm/arch/64/mode/hardware.h:
```diff
@@ -173,42 +240,50 @@
+
+#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
+/* The log buffer is placed before the device region using a 2MiB page. */
+#define KS_LOG_BASE UL_CONST(0x000000ffffc00000)  /* 2^48 - 2 * 2^21 */
```
apparently I renamed this from KS_LOG_PPTR which breaks benchmark builds. I should fix this before this is merged.

midnightveil committed:
aarch64: use fixed virtual addr for kernel image

This is a precursor to some other boot protocol changes as a part of
the multikernel work where it is useful to be able to load the kernel
at different physical addresses using the same kernel image.

Previously, the kernel would run in the kernel physical memory window
at a device-specific virtual address. Now, the virtual address is
constant, and so the virtual-physical offset is different per
platform, but still constant.

Signed-off-by: Julia Vassiliki <julia.vassiliki@unsw.edu.au>

midnightveil committed:
cleanup: use kernel_mapping_pds more

This makes the code look cleaner.

Signed-off-by: Julia Vassiliki <julia.vassiliki@unsw.edu.au>

midnightveil committed:
cleanup: convert asserts to compile_asserts

asserts() will silently fail as this is pre-vspace setup

Signed-off-by: Julia Vassiliki <julia.vassiliki@unsw.edu.au>

midnightveil committed:
cleanup: use function for mapping physical window

Diff looks a bit weird but the code is nicer.

Signed-off-by: Julia Vassiliki <julia.vassiliki@unsw.edu.au>

midnightveil committed:
aarch64: map phys window XN now ELF is separate

Because the ELF no longer lives in the physical memory
window, we can map the ELF is executable but the physical
window as execute-never.

Signed-off-by: Julia Vassiliki <julia.vassiliki@unsw.edu.au>

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,6 +233,87 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+
+    paddr_t paddr = KERNEL_ELF_PADDR_BASE;
+    pptr_t vaddr = KERNEL_ELF_BASE;
+    for (word_t pd = pd_start; pd < pd_end; pd++) {
```
I think this should be `pd <= pd_end`, see above.

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -284,6 +365,13 @@ BOOT_CODE void map_kernel_window(void)
+    pte_t *kernel_mapping_pds = armKSGlobalKernelPDs[kernel_mappings_pud_idx];
+
+    /* map the kernel ELF */
+    map_kernel_elf_image(kernel_mapping_pds);
```
You should call it as `map_kernel_elf_image(armKSGlobalKernelPDs[kernel_mappings_pud_idx])` to conform to the calling convention you specified.

midnightveil replied:
I don't exactly understand why you say this currently doesn't match the "calling convention"? The types match and it's the same thing.?

midnightveil replied:
That's only because there isn't a nice way to type `pte_t kernel_mapping_pds[]` as a local variable in C, though. They're the same type.

Indanz reviewed (commented):
Looks almost ready to me.

Indanz reviewed (commented):
Previous review was for the second commit only, this is for the rest (ignoring the doc one).

Still not sure about commit "nfc/cleanup: fill physical window using array indices", my feeling is that it only makes sense if you also switch to huge pages.

midnightveil commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,6 +233,87 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+    assert(KERNEL_ELF_TOP <= KDEV_BASE);
+
+    const word_t pd_start = GET_KPT_INDEX(KERNEL_ELF_BASE, KLVL_FRM_ARM_PT_LVL(2));
+    const word_t pd_end   = GET_KPT_INDEX(KERNEL_ELF_TOP, KLVL_FRM_ARM_PT_LVL(2));
```
If I did KERNEL_ELF_TOP - 1 as well that would work regardless of the alignment then yes?

Is there a better way to do that? It feels somewhat ugly.

(If it did get 2MiB align we make an extra unnecessary 2MiB. Maybe that's fine?)

Indanz replied:
Yes, strictly speaking also doing KERNEL_ELF_TOP - 1 is more correct (and then -1 on the same line). But the chance of ki_end falling exactly on a 2MB boundary is miniscule, and what's the harm? (Famous last words...)

No, making the binary 2MB bigger is not fine.

midnightveil replied:
> chance of ki_end falling exactly on a 2MB boundary is miniscule, a

well, say, for instance, someone editing the linker script such that the kernel takes up exactly one 2MiB page :)

What do you mean by "then -1 on the same line"?

midnightveil commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -357,7 +357,7 @@ BOOT_CODE void map_kernel_physical_window(void)
+                                                                                                                        1, // XN
 #else
                                                                                                                         1, // UXN
 #endif
```
In what other commit? I'm confused, this is a standalone commit, and I'm not sure the relevance of hyp vs non-hyp here.

midnightveil replied:
Because UXN is userspace execute never, I don't think we have control over the kernel execute never bits in hyp mode, i.e. it's always executable at the kernel level (IIRC).

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,6 +233,87 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+ */
+compile_assert(elf_in_kernel_mappings, GET_KPT_INDEX(KERNEL_ELF_BASE, KLVL_FRM_ARM_PT_LVL(1)) == BIT(PT_INDEX_BITS) - 1);
+compile_assert(kdev_in_kernel_mappings, GET_KPT_INDEX(KDEV_BASE, KLVL_FRM_ARM_PT_LVL(1))      == BIT(PT_INDEX_BITS) - 1);
+compile_assert(log_buf_in_kernel_mappings, GET_KPT_INDEX(KS_LOG_BASE, KLVL_FRM_ARM_PT_LVL(1)) == BIT(PT_INDEX_BITS) - 1);
```
If it's uglier than `BIT(PT_INDEX_BITS) - 1`, your define's name is definitely too long.

midnightveil replied:
As in, having to do a `#define` because C doesn't support proper constants.

midnightveil commented:
> I think actually after the rest of the cleanup I think it seems alright. Maybe I'll just explain it with some comments a bit more and drop the change.

I've decided to do this.

Here's that patch for posterity.

[0001-nfc-cleanup-fill-physical-window-using-array-indices.patch](https://github.com/user-attachments/files/23466612/0001-nfc-cleanup-fill-physical-window-using-array-indices.patch)

JDuchniewicz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,28 +233,116 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
-    armKSGlobalKernelPGD[GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0))] = pte_pte_table_new(
-                                                                                 addrFromKPPtr(armKSGlobalKernelPUD));
+#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
+/* verify that the physical memory window as at the second entry of the PGD */
```
```suggestion
/* verify that the physical memory window is the second entry of the PGD */
```

If I understand correctly ofc.

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,28 +233,116 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
 #ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-    /* verify that the kernel window as at the second entry of the PGD */
-    assert(GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0)) == 1);
+                                     /* XN */ 0,
```
I think we can set this to 1 too, as it controls execution at EL1 and EL0, not at EL2.

midnightveil commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,28 +233,116 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
 #else
-    /* verify that the kernel window as at the last entry of the PGD */
-    assert(GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0)) == BIT(PT_INDEX_BITS) - 1);
+                                     /* UXN */ 1,
```
Previously discussed: https://github.com/seL4/seL4/pull/1516#discussion_r2496647326

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,28 +233,116 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
-    armKSGlobalKernelPGD[GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0))] = pte_pte_table_new(
-                                                                                 addrFromKPPtr(armKSGlobalKernelPUD));
+#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
+/* verify that the physical memory window is at the second entry of the PGD */
```
Yeah, I saw that commit, but it doesn't make sense.

midnightveil commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,28 +233,116 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+                                 );
+
+        paddr += BIT(seL4_LargePageBits);
+        vaddr += BIT(seL4_LargePageBits);
```
These paddr/vaddr variables look like rebase remannds I need to fix.

lsf37 replied:
clang is complaining about `vaddr` being unused (correctly), although apparently only in release builds. `paddr` is used in the statement above.

midnightveil replied:
Right, this actually isn't a rebase remmant. I'm not sure why this wasn't an issue before; but it's because it was only in the assert. I did do full hw-test runs before so not sure why it only failed recently, but anyway.

I've added the `UNUSED` attribute.

lsf37 commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,28 +233,116 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+compile_assert(elf_in_kernel_mappings, GET_KPT_INDEX(KERNEL_ELF_BASE,
+                                                     KLVL_FRM_ARM_PT_LVL(1)) == BIT(PT_INDEX_BITS) - 1);
+compile_assert(kdev_in_kernel_mappings, GET_KPT_INDEX(KDEV_BASE,
+                                                      KLVL_FRM_ARM_PT_LVL(1))      == BIT(PT_INDEX_BITS) - 1);
```
```suggestion
                                                      KLVL_FRM_ARM_PT_LVL(1)) == BIT(PT_INDEX_BITS) - 1);
```

Indanz replied:
That still does the weird line split halfway GET_KPT_INDEX, making it hard to read. I'd go for an earlier new line and do something like:

```
compile_assert(kdev_in_kernel_mappings,
               GET_KPT_INDEX(KDEV_BASE, KLVL_FRM_ARM_PT_LVL(1)) == BIT(PT_INDEX_BITS) - 1);
```
But if we want to check that they are all equal, we could check that assumption directly with:

```
compile_assert(elf_and_kdev_in_same_pud,
               GET_KPT_INDEX(KERNEL_ELF_BASE, KLVL_FRM_ARM_PT_LVL(1)) ==
               GET_KPT_INDEX(KDEV_BASE,       KLVL_FRM_ARM_PT_LVL(1)));
```

midnightveil replied:
We'd still need another assert since there's 3 things in that top placement. Plus a few other spots assume `BIT(PT_INDEX_BITS) - 1` in this code. In theory this could also affect the loader, but so could all of this. (both Microkit and elfloader will continue working)

(applied Gerwin's suggestion)

lsf37 commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,28 +233,116 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+ *  the current ARMv8 A-Profile specifications no longer contain this wording,
+ *  and instead specify the loss of coherency and other behaviours in B2.11.
+ *
+ */
```
```suggestion
 */
```

lsf37 commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,28 +233,116 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+ *
+ *  Hence, the kernel should never access kernel ELF memory through the physical
+ *  memory window, i.e., never modify the result of `ptrFromPAddr()` when the
+ *  physical address might lie within the kernel ELF. (Reads are fine if they
```
```suggestion
 *  physical address lies within the kernel ELF. (Reads are fine if they
```

We should know whether it does or not, and we do know for all accesses in the kernel which region they are in.

lsf37 reviewed (approved):
Good to go from the verification side one the proofs in https://github.com/seL4/l4v/pull/1031 have passed.

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -233,6 +233,88 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+
+    const word_t pd_start = GET_KPT_INDEX(KERNEL_ELF_BASE, KLVL_FRM_ARM_PT_LVL(2));
+    /* handle the fact that ELF_TOP is exclusive, not inclusive */
+    const word_t pd_end   = GET_KPT_INDEX(KERNEL_ELF_TOP - 1, KLVL_FRM_ARM_PT_LVL(2)) + 1;
```
Really worth commenting here that `pd_end` is exclusive (because of the +1). That's very easy to miss, especially with the comment seemingly implying that we want an inclusive value.

midnightveil replied:
Reworded.

Indanz commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -283,16 +371,30 @@ BOOT_CODE void map_kernel_window(void)
+{
+    /* place the PUD into the PGD */
+    armKSGlobalKernelPGD[GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0))]
+        = pte_pte_table_new(addrFromKPPtr(armKSGlobalKernelPUD));
```
```suggestion
    armKSGlobalKernelPGD[GET_KPT_INDEX(PPTR_BASE, KLVL_FRM_ARM_PT_LVL(0))] =
        pte_pte_table_new(addrFromKPPtr(armKSGlobalKernelPUD));
```
very unusual in C to put the "=" on the next line when you break it up.

(But this monstrosity can be easily fixed by splitting it up and using local vars for the indices.)

midnightveil replied:
Moved.

Indanz reviewed (approved):
Other than some nitpicks all seems fine.

midnightveil commented on src/arch/arm/64/kernel/vspace.c:
```diff
@@ -283,16 +371,30 @@ BOOT_CODE void map_kernel_window(void)
+    map_kernel_physical_window();
+
+    const word_t kernel_mappings_pud_idx = GET_KPT_INDEX(KERNEL_ELF_BASE, KLVL_FRM_ARM_PT_LVL(1));
+    pte_t *kernel_mapping_pds = armKSGlobalKernelPDs[kernel_mappings_pud_idx];
```
Renaming to `elf_pud_idx`.
